packages/data-context/src/util/weightedChoice.ts

```diff
@@ -0,0 +1,55 @@
+import _ from 'lodash'
+
+export type WeightedAlgorithm = {
+  pick: (values: any[]) => string
+}
+
+/**
+ * Randomly choose an index from an array based on weights
+ *
+ * Based on algorithm found here: https://dev.to/trekhleb/weighted-random-algorithm-in-javascript-1pdc
+ *
+ * @param weights array of numbered weights that correspond to the indexed values
+ * @param values array of values to choose from
+ */
+const weightedChoice = (weights: number[], values: any[]) => {
+  if (weights.length > 0 && values.length > 0 && weights.length !== values.length) {
+    throw new Error('The length of the weights and values must be the same and greater than zero')
+  }
+
+  const cumulativeWeights = _.reduce(weights, (result: number[], value: number) => {
+    if (result.length === 0) {
+      result.push(value)
+    } else {
+      // @ts-ignores
+      result.push(value + result[result.length - 1])
+    }
+
+    return result
+  }, [])
+
+  // @ts-ignores
+  const randomNumber = Math.random() * cumulativeWeights[cumulativeWeights.length - 1]
+
+  const choice = _.transform(cumulativeWeights, (result, value, index) => {
+    if (value >= randomNumber) {
+      result.chosenIndex = index
+    }
+
+    return result.chosenIndex === -1
+  }, { chosenIndex: -1 })
+
+  return values[choice.chosenIndex]
+}
```

Comment:
Any ideas for not needing the two `@ts-ignores` I added for referencing the last value in an array?

Reply:
Hmm maybe this:

Only line I am unsure about

```js
const randomNumber = Math.random() * (cumulativeWeights[cumulativeWeights.length - 1] ?? 1)
```

Should we error if `cumulativeWeights[cumulativeWeights.length - 1]` is null? Should be impossible?

```js
import _ from 'lodash'

export type WeightedAlgorithm = {
  pick: (values: string[]) => string
}

/**
 * Randomly choose an index from an array based on weights
 *
 * Based on algorithm found here: https://dev.to/trekhleb/weighted-random-algorithm-in-javascript-1pdc
 *
 * @param weights array of numbered weights that correspond to the indexed values
 * @param values array of values to choose from
 */
const weightedChoice = (weights: number[], values: any[]) => {
  if (weights.length === 0 || values.length === 0 || weights.length !== values.length) {
    throw new Error('The length of the weights and values must be the same and greater than zero')
  }

  const cumulativeWeights = weights.reduce<number[]>((acc, curr) => {
    if (acc.length === 0) {
      return [curr]
    } 

    const next = acc[acc.length - 1]

    if (!next) {
      return acc
    }

    return [...acc, next]
  }, [])

  const randomNumber = Math.random() * (cumulativeWeights[cumulativeWeights.length - 1] ?? 1)

  const choice = _.transform(cumulativeWeights, (result, value, index) => {
    if (value >= randomNumber) {
      result.chosenIndex = index
    }

    return result.chosenIndex === -1
  }, { chosenIndex: -1 })

  return values[choice.chosenIndex]
}

export const WEIGHTED = (weights: number[]): WeightedAlgorithm => {
  return {
    pick: (values: any[]): string => {
      return weightedChoice(weights, values)
    },
  }
}

export const WEIGHTED_EVEN = (values: any[]): WeightedAlgorithm => {
  return WEIGHTED(_.fill(Array(values.length), 1))
}
```